spinalcordtoolbox/reports/slice.py

```diff
@@ -444,3 +456,44 @@ def get_center(self, img_idx=-1):
         size_y = self.axial_dim(image)
         size_x = self.coronal_dim(image)
         return np.ones(dim) * size_x / 2, np.ones(dim) * size_y / 2
+
+    def mosaic(self, nb_column=0, dilation=(0, 0, 0)):
```

Comment:
Oh, I've also added this `dilation` parameter which we can use to dilate the bounding box, just in case we want to capture voxels outside of the SC segmentation bounding box, too.

The only issue here is that the `mosaic` method signatures don't match the base `Slice` class anymore. Is there a way to define the base `mosaic` abstract method to accommodate for different signatures?

Reply:
Several things:
- The first technical answer is that Python itself doesn't care about the method signature, and neither does the `@abc.abstractmethod` decorator (it only checks whether _something_ called `mosaic` is defined in the derived class before allowing instances to be created). But that's not a very satisfying answer.
- A second answer, which isn't actually applicable here, is that if you want a fully general method signature in Python, you can do this:

  ```py3
  class Foo:
      def bar(self, *args, **kwargs):
          ...
  ```
- A third answer, which is more satisfying and applicable here, is given by [Mypy's rules for deciding which methods have compatible type signatures](https://mypy.readthedocs.io/en/stable/class_basics.html#overriding-statically-typed-methods). Basically, they implement the idea that every valid method call to the parent class's method has to still be a valid method call to the child class's method (and not necessarily the reverse). In particular, this means that:
  - Positional arguments that appear in the parent should appear in the child, with the same names and in the same order. Arguments that are optional in the parent (that is, arguments with a default value) should also be optional in the child. But mandatory arguments in the parent can be optional in the child, and the child can have more optional arguments.
  - Keyword-only arguments in the parent should be accepted in the child.
  - For each argument accepted in the parent, its type should be a subtype of what the child accepts for that argument.
  - Going in the other direction, the _return_ type of the child should be a subtype of what the parent's return type.

  (Aside: this is why the "fully general" method signature from the second answer above isn't applicable. A child method which is "fully general" is compatible with _any_ parent method, but not the other way around. In fact, a parent method which is "fully general" is maximally restrictive on the signature of a child method that wants to override it.)
- More concretely, just looking at the three method signatures here, we could make them compatible by putting only the common elements in the parent signature, since the other elements are optional, like so:
  ```py3
  class Slice:
      def mosaic(self, nb_column): pass

  class Axial(Slice):
      def mosaic(self, nb_column=0, size=15, return_center=False): pass

  class Sagittal(Slice):
      def mosaic(self, nb_column=0, dilation=(0, 0, 0)): pass
  ```
- But, looking at how this method is actually called in our codebase, I'm only seeing two kinds of call:
  - `mosaic()`, with no arguments at all; and
  - `mosaic(return_center=True)`, which uses a named argument, and only gets called on instances of the `Axial` subclass.

  So, possibly we should simplify the signatures to:
  ```py3
  class Slice:
      def mosaic(self): pass

  class Axial(Slice):
      def mosaic(self, return_center=False): pass

  class Sagittal(Slice):
      def mosaic(self): pass
  ```

  (One of the calls to `mosaic(return_center=True)` is done by the `Slice.mosaics_through_time` method, which should probably be moved to `Axial.mosaics_through_time`, unless you want to add `return_center` functionality to `Sagittal.mosaic` as well.)

Reply:
Thank you for such a deep and thorough answer!

> * But, looking at how this method is actually called in our codebase, I'm only seeing two kinds of call:

This is a very good point. `nb_column`, `size`, and `dilation` are definitely speculative features? They assume that we may one day want to tweak how the axial/sagittal grids are displayed.

But, as you point out, so far we haven't needed to change these values, and thus they've gone unused. 

So, I would be OK with a follow-up PR that removes these extraneous arguments. (I suggest a separate PR as that way we have a clear place to view the precise changes, if/when we ever need to re-add these features).

Reply:
> (One of the calls to mosaic(return_center=True) is done by the Slice.mosaics_through_time method, which should probably be moved to Axial.mosaics_through_time, unless you want to add return_center functionality to Sagittal.mosaic as well.)

I absolutely agree here. `return_center` is definitely only relevant for axial slices.